controllers/loadtest_controller.go

```diff
@@ -175,131 +176,221 @@ func (r *LoadTestReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-
-			if updateErr := r.Status().Update(ctx, test); updateErr != nil {
-				logWithServer.Error(updateErr, "failed to update status after failure to create pod for server")
+		poolAvailabilities := make(map[string]int)
+		for pool, capacity := range poolCapacities {
+			poolAvailabilities[pool] = capacity
+		}
+		for _, pod := range pods.Items {
+			pool, ok := pod.Labels[config.PoolLabel]
+			if !ok {
+				log.Info("encountered a pod without a pool label", "pod", pod)
+				continue
+			}
+			if pod.Status.Phase != corev1.PodSucceeded && pod.Status.Phase != corev1.PodFailed {
+				poolAvailabilities[pool]--
 			}
-
-			return *result, err
 		}
-	}
-	for i := range missingPods.Clients {
-		logWithClient := log.WithValues("client", missingPods.Clients[i])
-
-		pod, err := builder.PodForClient(&missingPods.Clients[i])
-		if err != nil {
-			logWithClient.Error(err, "failed to construct a pod struct for supplied client struct")
 
-			test.Status.State = grpcv1.Errored
-			test.Status.Reason = grpcv1.ConfigurationError
-			test.Status.Message = fmt.Sprintf("failed to construct a pod for client at index %d: %v", i, err)
-
-			if updateErr := r.Status().Update(ctx, test); updateErr != nil {
-				logWithClient.Error(updateErr, "failed to update status after failure to construct a pod for client")
+		adjustAvailabilityForDefaults := func(defaultPoolKey, defaultPoolName string) {
+			if c, ok := missingPods.NodeCountByPool[defaultPoolKey]; ok {
+				missingPods.NodeCountByPool[defaultPoolName] += c
 			}
-
-			return ctrl.Result{}, err
+			delete(missingPods.NodeCountByPool, defaultPoolKey)
 		}
+		adjustAvailabilityForDefaults(status.DefaultClientPool, defaultClientPool)
+		adjustAvailabilityForDefaults(status.DefaultDriverPool, defaultDriverPool)
+		adjustAvailabilityForDefaults(status.DefaultServerPool, defaultServerPool)
+
+		for pool, requiredNodeCount := range missingPods.NodeCountByPool {
+			availableNodeCount, ok := poolAvailabilities[pool]
+			if !ok {
+				log.Error(errNonexistentPool, "requested pool does not exist and cannot be considered when scheduling", "requestedPool", pool)
+				test.Status.State = grpcv1.Errored
+				test.Status.Reason = grpcv1.PoolError
+				test.Status.Message = fmt.Sprintf("requested pool %q does not exist", pool)
+				if updateErr := r.Status().Update(ctx, test); updateErr != nil {
+					log.Error(updateErr, "failed to update status after failure due to requesting nodes from a nonexistent pool")
+				}
+				return ctrl.Result{}, errNonexistentPool
+			}
 
-		result, err := createPod(pod)
-		if result != nil && !kerrors.IsAlreadyExists(err) {
-			logWithClient.Error(err, "failed to create pod for client")
+			if requiredNodeCount > availableNodeCount {
+				log.Info("cannot schedule test: inadequate availability for pool", "pool", pool, "requiredNodeCount", requiredNodeCount, "availableNodeCount", availableNodeCount)
+				return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
```

Comment:
Based on our current experience, this ctrl.Result{RequeueAfter: 5 * time.Second} might not happen, just a thing to watch out...

Reply:
I witnessed this only after a failure for >30min to schedule. It is something to lookout for. We can consider two things: 

1. We can adjust the Resync ratio (though not recommended) to re-reconcile everything in the cache at a certain interval as a backup.

1. We could introduce some component in a separate go routine or separate deployment that triggers reconciles after a long period of idle time by bumping the generation number.

Reply:
I am reading controller-runtime code base as a hobby, lol, learned a lot. I am suspecting if any reconcile is happening due to ctrl.Result{Requeue: true} or an error returned (not reconcile due to a resource update ) before RequeueAfter time is up, the RequeueAfter could be requested. Still investigating....I think Solly has the point of something is update the lock, although I am not so sure that is what is happening but I think it is a direction worth exploring